File: cpp/organism/plant/species/Belladonna.cpp
//
// Created by fgrel on 18.04.2025.
//

#include "Belladonna.h"

void Belladonna::reproduce() {
  Point newPosition = getNewPosition();
  if (newPosition == position)
    return; // Ensure a new position is valid

  world.killPlantsOnPosition(newPosition);
  world.queueOrganismAddition(new Belladonna(newPosition, world));
}

double Belladonna::getSpreadProbability() const {
  return spreadProbability;
}

bool Belladonna::collision(Organism &other) {
  world.getLogger().logEvent("Belladonna has killed  " +
                              other.getSymbol() + ".");
  other.die();
  return false;
}

Organism* Belladonna::clone() const {
    return new Belladonna(*this); // Tworzy kopię obiektu za pomocą konstruktora kopiującego
}

File: java/src/organism/plant/species/Belladonna.java
package organism.plant.species;

import constants.Constants;
import logger.EventLogger;
import organism.Organism;
import organism.plant.Plant;
import world.World;

public class Belladonna extends Plant {

    public Belladonna(int x, int y, World world) {
        super(x, y, Constants.Plant.Belladonna.Symbol, Constants.Plant.Belladonna.Strength, Constants.Plant.Initiative, world);
    }

    @Override
    protected void reproduce() {
        int[] newPosition = getNewPosition();
        if (newPosition[0] == x && newPosition[1] == y) return;
        world.addOrganism(new Belladonna(newPosition[0], newPosition[1], world));
        EventLogger.getInstance().log("Belladonna spread to (" + newPosition[0] + ", " + newPosition[1] + ")");
    }

    @Override
    protected double getSpreadProbability() {
        return Constants.Plant.Belladonna.SpreadProbability;
    }

    @Override
    public boolean collision(Organism other) {
        EventLogger.getInstance().log("Belladonna has killed " + other.getClass().getSimpleName() + " at (" + x + ", " + y + ")");
        other.die();
        return false;
    }
}